src/components/deck-gl-wrapper/custom-tile-3d-layer.ts

```diff
@@ -0,0 +1,572 @@
+    }
+
+    const options = { loadOptions: { ...loadOptions } };
+    if (loader.preload) {
+      const preloadOptions = await loader.preload(tilesetUrl, loadOptions);
+
+      if (preloadOptions.headers) {
+        options.loadOptions.fetch = {
+          ...options.loadOptions.fetch,
+          headers: preloadOptions.headers,
+        };
+      }
+      Object.assign(options, preloadOptions);
+    }
+    const tilesetJson = await load(tilesetUrl, loader, options.loadOptions);
+
+    const tileset3d = new Tileset3D(tilesetJson, {
+      onTileLoad: this._onTileLoad.bind(this),
+      onTileUnload: this._onTileUnload.bind(this),
+      onTileError: this.props.onTileError,
+      onTraversalComplete: this._onTraversalComplete.bind(this),
+      ...options,
+    });
+
+    this.setState({
+      tileset3d,
+      layerMap: {},
+    });
+
+    this._updateTileset(this.state.activeViewports);
+    this.props.onTilesetLoad(tileset3d);
+  }
+
+  private _onTileLoad(tileHeader: Tile3D): void {
+    const { lastUpdatedViewports } = this.state;
+
+    if (tileHeader.type === TILE_TYPE.MESH) {
+      tileHeader.userData.originalColorsAttributes = {
+        ...tileHeader.content.attributes.colors,
+        value: tileHeader.content.attributes.colors.value.slice(),
+      };
+      this._colorizeTile(tileHeader);
+    }
+
+    this.props.onTileLoad(tileHeader);
+    if (!this.state.colorsByAttribute) {
+      this._updateTileset(lastUpdatedViewports);
+      this.setNeedsUpdate();
+    }
+  }
+
+  private _onTraversalComplete(selectedTiles: Tile3D[]): Tile3D[] {
+    if (
+      this.state.isCustomColors &&
+      selectedTiles[0]?.type === TILE_TYPE.MESH
+    ) {
+      selectedTiles.forEach((tile) => {
+        if (
+          tile.content &&
+          tile.userData.customColors !== this.state.colorsByAttribute &&
+          tile.userData.originalColorsAttributes
+        ) {
+          tile.content.attributes.colors.value =
+            tile.userData.originalColorsAttributes.value.slice();
+          this._colorizeTile(tile);
+        }
+      });
+    }
+    return selectedTiles;
+  }
+
+  private _onTileUnload(tileHeader: Tile3D): void {
+    // Was cleaned up from tileset cache. We no longer need to track it.
+    delete this.state.layerMap[tileHeader.id];
+    this.props.onTileUnload(tileHeader);
+  }
+
+  private _updateTileset(
+    viewports: { [viewportId: string]: Viewport } | null
+  ): void {
+    if (!viewports) {
+      return;
+    }
+    const { tileset3d } = this.state;
+    const { timeline } = this.context;
+    const viewportsNumber = Object.keys(viewports).length;
+    if (!timeline || !viewportsNumber || !tileset3d) {
+      return;
+    }
+    tileset3d.selectTiles(Object.values(viewports)).then((frameNumber) => {
+      const tilesetChanged = this.state.frameNumber !== frameNumber;
+      if (tilesetChanged) {
+        this.setState({ frameNumber });
+      }
+    });
+  }
+
+  private _colorizeTile(tile: Tile3D): void {
```

Comment:
This function might be done as a method in `Tile3D` class. I suspect we can move the most of logic to loaders.gl. It is the preferrable approach because we need to change only data and that is what loaders.gl about.